firefox-ios/Client/Frontend/Browser/WebView/PullRefreshView.swift

```diff
@@ -0,0 +1,323 @@
+                                                                                                            y: shrinkFactor)
+            self.progressContainerView.backgroundColor = self.currentTheme?.colors.layer4
+        }, completion: { _ in
+            UIImpactFeedbackGenerator(style: .heavy).impactOccurred()
+        })
+    }
+
+    private func restoreBackgroundProgressViewIfNeeded() {
+        guard refreshIconHasFocus else { return }
+        refreshIconHasFocus = false
+        UIView.animate(withDuration: 0.3) {
+            self.progressContainerView.transform = .identity
+            self.progressContainerView.backgroundColor = .clear
+        }
+    }
+
+    private func showEasterEgg() {
+        guard let easterEggGif else { return }
+        easterEggGif.isHidden = false
+        let angle = atan2(easterEggGif.transform.b, easterEggGif.transform.a)
+        UIView.animate(withDuration: 0.3) {
+            easterEggGif.transform = .identity.rotated(by: angle)
+        }
+    }
+
+    private func computeShrinkingFactor() -> CGFloat {
+        guard let scrollView else { return 1.0 }
+        let minDimension = min(scrollView.frame.height, scrollView.frame.width)
+        if minDimension / 4.0 > UX.blinkProgressViewStandardThreshold {
+            return 1
+        }
+        return 0.8
+    }
+
+    func stopObservingContentScroll() {
+        obeserveTicket?.invalidate()
+    }
+
+    func updateEasterEggForOrientationChange(isPotrait: Bool) {
+        guard let easterEggGif else { return }
+        easterEggGif.removeFromSuperview()
+        insertSubview(easterEggGif, at: 0)
+        let shrinkFactor = computeShrinkingFactor()
+        let easterEggSize = UX.easterEggSize.applying(CGAffineTransform(scaleX: shrinkFactor, y: shrinkFactor))
+        let layoutBuilder = EasterEggViewLayoutBuilder(easterEggSize: easterEggSize)
+        layoutBuilder.layoutEasterEggView(easterEggGif, superview: self, isPotrait: isPotrait, isIpad: isIpad)
+    }
+
+    // MARK: - ThemeApplicable
+
+    func applyTheme(theme: any Theme) {
+        currentTheme = theme
+        backgroundColor = theme.colors.layer1
+        progressView.tintColor = theme.colors.iconPrimary
+    }
+
+    private func loadGifFromBundle(named name: String) -> UIImageView? {
+        guard let gifPath = Bundle.main.path(forResource: name, ofType: "gif"),
+              let gifData = NSData(contentsOfFile: gifPath) as Data?,
+              let source = CGImageSourceCreateWithData(gifData as CFData, nil) else {
+            return nil
+        }
+
+        var frames: [UIImage] = []
+        let frameCount = CGImageSourceGetCount(source)
+
+        for i in 0..<frameCount {
+            if let cgImage = CGImageSourceCreateImageAtIndex(source, i, nil) {
+                frames.append(UIImage(cgImage: cgImage))
+            }
+        }
+
+        let animatedImage = UIImage.animatedImage(with: frames, duration: Double(frameCount) * 0.1)
```

Comment:
This is really helpful for loading gifs! I wonder if we should move the part for getting an animated UIImage to `UIImageExtensions.swift` so we can reuse this elsewhere...?

Reply:
Yes true i totally agree i'll move it there 😃